controllers/watcherdecisionengine_controller.go

```diff
@@ -175,22 +217,48 @@ func (r *WatcherDecisionEngineReconciler) Reconcile(ctx context.Context, req ctr
 		}
 	}
 
-	err = r.generateServiceConfigs(ctx, instance, secret, memcached, helper, &configVars)
+	err = r.generateServiceConfigs(ctx, instance, secret, prometheusSecret, memcached, helper, &configVars)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	Log.Info(fmt.Sprintf("[DecisionEngine] Getting input hash '%s'", instance.Name))
+	//
+	// create hash over all the different input resources to identify if any those changed
+	// and a restart/recreate is required.
+	//
+
+	inputHash, hashChanged, errorHash := r.createHashOfInputHashes(ctx, instance, configVars)
+	if errorHash != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.ServiceConfigReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.ServiceConfigReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	} else if hashChanged {
+		// Hash changed and instance status should be updated (which will be done by main defer func),
+		// so we need to return and reconcile again
+		return ctrl.Result{}, nil
+	}
+
 	instance.Status.Conditions.MarkTrue(condition.ServiceConfigReadyCondition, condition.ServiceConfigReadyMessage)
 
+	result, err = r.ensureDeployment(ctx, helper, instance, prometheusSecret, inputHash)
+	if err != nil {
```

Comment:
we should check for result here, as `ensureDeployment` can return a result diferent than `ctrl.Result{}`, a condition like `(err != nil || result != ctrl.Result{})` would be better ( I know that the error come from the API, I'll fix that in my next PR).

Reply:
I think we don't need it. See the coment in ensureDeployment:

// It is OK to return success as we are watching for StatefulSet changes

My understanding is that that's a case where we are still setting the DeploymentReady condition to False but we don't need to request for a reconcile by returning error because a reconcile will automatically be created when the StatefulSet status changes.

Reply:
FYI i added a followup comment at https://github.com/openstack-k8s-operators/watcher-operator/commit/aecb03d9d2415a26ad158d9a22310bfb29d1946c#r152165005 after our discussion earlier